### PR TITLE
Add Update Email View

### PR DESCRIPTION
**Pull request by raisedadead**
This commit:
- Displays the user's email that we have on records
- Adds a button in the settings to update email
- Adds a form view to update the email
- Fixes CSS for the Email icons and the email form ~ Credits to @hallaathrad
## **Settings Page - With No-Email**

screen shot 2016-04-24 at 4 41 48 am
## **Sucess Message**

screen shot 2016-04-24 at 4 45 02 am

---

---
## **Settings Page - With Email**

screen shot 2016-04-24 at 4 46 44 am

---

### CODE REVIEW

**Comment by raisedadead** [comment]
Addresses some parts of #8167 .

File: server/boot/user.js [hunk 1]
```diff
@@ -235,6 +238,39 @@ module.exports = function(app) {
...
+  function getUpdateEmail(req, res) {
+    if (!req.user) {
+      return res.redirect('/');
+    }
+    return res.render('account/email-update', {
+      title: 'Update your Email'
+    });
+  }
+
+  function postUpdateEmail(req, res, next) {
```
Comment (BerkeleyTrue) [inline server/boot/user.js]: These api endpoints should be added using loopbacks build in `remoteMethods`
see https://docs.strongloop.com/display/APIC/Remote+methods
and https://github.com/FreeCodeCamp/FreeCodeCamp/blob/staging/common/models/user.js#L265

File: common/models/user.js [hunk 2]
```diff
@@ -306,6 +306,88 @@ module.exports = function(User) {
...
+
+  User.updateEmail = function updateEmail(email, req, res, cb) {
+    if (!email) {
+      return nextTick(() => {
+        cb(new TypeError(
+            `email should be valid, got ${ email }`
+        ));
+      });
+    }
+    const userId = req.user.id;
```
Comment (BerkeleyTrue) [inline common/models/user.js]: This won't be available during api requests.

File: common/models/user.js [hunk 3]
```diff
@@ -306,6 +306,88 @@ module.exports = function(User) {
...
+        console.error(err);
+        ctx.req.flash('error', {
+          msg: 'Oops, something went wrong, please try again later'
+        });
+        return ctx.res.redirect('/settings');
+      });
+  });
+
+  User.updateEmail = function updateEmail(email, req, res, cb) {
+    if (!email) {
```
Comment (BerkeleyTrue) [inline common/models/user.js]: This should be an instance method in order to ensure you have the correct user.

File: common/models/user.js [hunk 4]
```diff
@@ -306,6 +306,88 @@ module.exports = function(User) {
...
+        ));
+      });
+    }
+    const userId = req.user.id;
+    return User.findById(userId, function(err, user){
+      user.updateAttribute('email',email,function(err){
+        if (err) { return next(err); }
+        debug('updated email for user');
+        req.flash('info', { msg: 'You\'ve successfully updated your email.' });
+        return res.redirect('/settings');
```
Comment (BerkeleyTrue) [inline common/models/user.js]: These details should be handled by the client, so that the API endpoint can be used in different places.

**Comment by BerkeleyTrue** [comment]
@raisedadead Looks like we are getting close. Great work so far. Let me know if you time to pair.

**Comment by raisedadead** [comment]
@BerkeleyTrue I have gotten rid of the `beforeRemote` hook, turns out it is never triggered for custom instance methods.

Had my head spinning for a day with that.

Now the validations are integrated with the User instance remote method, i.e `User.prototype.updateEmail`.

Now I just need a way to figure out how to send in authenticated request for the API call. 

Added inline comments for you reference, when you got time lets pair once again, to get this done.

File: server/views/account/email-update.jade [hunk 5]
```diff
@@ -0,0 +1,14 @@
+extends ../layout
+block content
+  h2.text-center Update your email address here:
+  form.form-horizontal.email-update(method='POST', action='/api/users/#{user.id}/email-update', name="updateEmailForm")
```
Comment (raisedadead) [inline server/views/account/email-update.jade]: `/api/users/#{user.id}/email-update`

This MUST be changed or anyone can simply call and change anyone else's data via a brute-force.

Comment (raisedadead) [inline server/views/account/email-update.jade]: /cc @BerkeleyTrue

Comment (BerkeleyTrue) [inline server/views/account/email-update.jade]: This is fine.

File: common/models/user.js [hunk 6]
```diff
@@ -306,6 +306,50 @@ module.exports = function(User) {
     }
   );
 
+  User.prototype.updateEmail = function updateEmail(email) {
+    return User.doesExist(null, email)
+      .then(exists => {
+        if (!exists) {
+          return this.update$({ email }).toPromise();
+        }
+        if (this.email && this.email === email) {
```
Comment (BerkeleyTrue) [inline common/models/user.js]: This check should be done first, then the check for emails existance

File: common/models/user.json [hunk 7]
```diff
@@ -235,6 +235,13 @@
       "principalId": "$everyone",
       "permission": "ALLOW",
       "property": "giveBrowniePoints"
+    },
+    {
+      "accessType": "EXECUTE",
+      "principalType": "ROLE",
+      "principalId": "$everyone",
```
Comment (BerkeleyTrue) [inline common/models/user.json]: This should be restricted to '$owner'

**Comment by BerkeleyTrue** [comment]
I've pulled this down locally. If I purposefully fail the email update it logs me out. Are you seeing the same?

File: common/models/user.js (hunk 6 above)
Comment (raisedadead) [inline common/models/user.js]: Agreed. Will make the check go first.

File: common/models/user.json (hunk 7 above)
Comment (raisedadead) [inline common/models/user.json]: Yup.

**Comment by raisedadead** [comment]
Oh I didn't check this. I'll check.

**Comment by raisedadead** [comment]
@BerkeleyTrue Please have a look.
@hallaathrad I would love some feedback from you as well for the overall UX, I know I can improve more on that jade code, but apart from that this PR is good for a test now.

File: server/boot/user.js [hunk 8]
```diff
@@ -226,6 +227,7 @@ module.exports = function(app) {
     res.redirect('/');
   }
 
+
```
Comment (theflametrooper) [inline server/boot/user.js]: This line should be removed :wink:

**Pull request merged**